### ISSUE TITLE
extends error!

### ISSUE THREAD

**Issue by imfox** [Perryvw/TypescriptToLua#142]
```typescript
class a { }
class b extends a { constructor() { super(); } }
class c extends b { constructor() { super(); } }
new c();
```
Will cause a stack overflow

**Comment by imfox** [comment]
```lua
a = a or {}
a.__index = a
function a.new(construct, ...)
    local instance = setmetatable({}, a)
    if construct and a.constructor then a.constructor(instance, ...) end
    return instance
end
function a.constructor(self)
end
b = b or a.new()
b.__index = b
b.__base = a
function b.new(construct, ...)
    local instance = setmetatable({}, b)
    if construct and b.constructor then b.constructor(instance, ...) end
    return instance
end
function b.constructor(self)
    self.__base.constructor(self)
end
c = c or b.new()
c.__index = c
c.__base = b
function c.new(construct, ...)
    local instance = setmetatable({}, c)
    if construct and c.constructor then c.constructor(instance, ...) end
    return instance
end
function c.constructor(self)
    self.__base.constructor(self)
end
c.new(true)
```
I think it can be changed to this
```lua
a = a or {}
a.__index = a
function a.new(construct, ...)
    local instance = setmetatable({}, a)
    if construct and a.constructor then a.constructor(instance, ...) end
    return instance
end
function a.constructor(self)
end
b = b or a.new()
b.__index = b
b.__base = a
function b.new(construct, ...)
    local instance = setmetatable({}, b)
    if construct and b.constructor then b.constructor(instance, ...) end
    return instance
end
function b.constructor(self)
    b.__base.constructor(self) -- here
end
c = c or b.new()
c.__index = c
c.__base = b
function c.new(construct, ...)
    local instance = setmetatable({}, c)
    if construct and c.constructor then c.constructor(instance, ...) end
    return instance
end
function c.constructor(self)
    c.__base.constructor(self) -- here
end
c.new(true)
```

**Comment by imfox** [comment]
When calling the superclass method, 
Maybe it should translate to
```lua
self.__base.xxxx(self);
```

this is wrong
```lua
self.__base:xxxx()
```

It's better to be
```lua
className.__base.xxxx(self)
```

**Comment by Perryvw** [comment]
It works like that last case now.

**Comment by imfox** [comment]
```typescript
class a {
    x = 1;
    constructor() {
        this.x = 2;
    }
}
new a();
```
translate to
```lua
a = a or {}
a.__index = a
function a.new(construct, ...)
    local instance = setmetatable({}, a)
    if construct and a.constructor then a.constructor(instance, ...) end
    instance.x = 1
    return instance
end
function a.constructor(self)
    self.x = 2
end
a.new(true)
```
The value of the x is covered
Maybe it should be before constructor statement

**Comment by Perryvw** [comment]
Fixed that and added a test for it in 452aab917677608231b66665f8057244107af98d, good catch.
